rust/src/ldap/detect.rs

```diff
@@ -407,6 +409,64 @@ unsafe extern "C" fn ldap_detect_responses_result_code_free(_de: *mut c_void, ct
     std::mem::drop(Box::from_raw(ctx));
 }
 
+unsafe extern "C" fn ldap_detect_responses_error_msg_setup(
+    de: *mut c_void, s: *mut c_void, _raw: *const std::os::raw::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_LDAP) != 0 {
+        return -1;
+    }
+    if DetectBufferSetActiveList(de, s, G_LDAP_RESPONSES_ERROR_MSG_BUFFER_ID) < 0 {
+        return -1;
+    }
+    return 0;
+}
+
+unsafe extern "C" fn ldap_detect_responses_error_msg_get_data(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int, local_id: u32,
+) -> *mut c_void {
+    return DetectHelperGetMultiData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        local_id,
+        ldap_tx_get_responses_error_msg,
+    );
+}
+
+unsafe extern "C" fn ldap_tx_get_responses_error_msg(
+    tx: *const c_void, _flags: u8, local_id: u32, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, LdapTransaction);
+
+    if local_id as usize >= tx.responses.len() {
+        return false;
+    }
+    *buffer = std::ptr::null();
+    *buffer_len = 0;
+
+    let response = &tx.responses[local_id as usize];
+    // We expect every response in one tx to be the same protocol_op
+    let str_buffer: &str = match &response.protocol_op {
+        ProtocolOp::BindResponse(req) => req.result.diagnostic_message.0.as_str(),
+        ProtocolOp::SearchResultDone(req) => req.diagnostic_message.0.as_str(),
+        ProtocolOp::ModifyResponse(req) => req.result.diagnostic_message.0.as_str(),
+        ProtocolOp::AddResponse(req) => req.diagnostic_message.0.as_str(),
+        ProtocolOp::DelResponse(req) => req.diagnostic_message.0.as_str(),
+        ProtocolOp::ModDnResponse(req) => req.diagnostic_message.0.as_str(),
+        ProtocolOp::CompareResponse(req) => req.diagnostic_message.0.as_str(),
+        ProtocolOp::ExtendedResponse(req) => req.result.diagnostic_message.0.as_str(),
+        _ => return false,
```

Comment:
I think we need to improve our loop and not return false in this case :

What about `SearchResultEntry` ?

If we have a tx with 2 responses : one SearchResultEntry and one SearchResultDone, we will not test the SearchResultDone as we will return false for the SearchResultEntry which gets interpreted as do not test further.

One easy workaround may be to return true with an buffer (and document it)

Also I would like rust compiler to help us catch that we are complete for all responses operations in these switches